File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string str) {
        stack<int> st;
        int ans = 0, n = str.length();
        for(int i = 0; i<n; i++){
            char c = str[i];
            if( c == '(') st.push(i);
            else{
                if(!st.empty() && str[st.top()] == '(') st.pop();
                else st.push(i);
            }
        }
        if(st.empty()) return str.length();
        else{
            int l = 0, r = str.length()-1;
            while(!st.empty()){
                l = st.top(); st.pop();
                ans = fmax(ans, r - l);
                r = l-1;
            }
            ans = fmax(ans, r+1);
        }
        return ans;
    }
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> st = new Stack<>();
        int longest = 0, n = s.length();
        for(int i = 0; i<n; i++){
            if(s.charAt(i) == '(') st.push(i);
            else{
                if(!st.isEmpty() && s.charAt(st.peek()) == '(') st.pop();
                else st.push(i);
            }
        }
        if(st.isEmpty()) return n;
        else{
            int l = 0, r = n - 1;
            while(!st.isEmpty()){
                l = st.pop();
                longest = Math.max(longest, r - l);
                r = l-1;
            }
            longest = Math.max(longest, r+1);
        }
        return longest;
    }
}